en/dev_log/ulog_file_format.md

```diff
@@ -41,233 +48,318 @@ The file consists of three sections:
+  struct message_header_s;
+  uint8_t compat_flags[8];
+  uint8_t incompat_flags[8];
+  uint64_t appended_offsets[3]; ///< file offset(s) for appended data if appending bit is set
+};
+```
+
+- `compat_flags`: compatible flag bits.
+  - `compat_flags[0]`, bit 0, *DEFAULT_PARAMETERS*: if set, the log contains parameter defaults (message 'Q').
+
+  The rest of the bits is currently not defined and all must be set to 0.
+  These bits can be used for future ULog changes that are compatible with existing parsers.
+  It means parsers can just ignore the bits if one of the unknown bits is set.
+- `incompat_flags`: incompatible flag bits.
+  The LSB bit of index 0 is set to one if the log contains appended data and at least one of the `appended_offsets` is non-zero.
+  All other bits are undefined and must be set to 0.
+  If a parser finds one of these bits set, it must refuse to parse the log.
+  This can be used to introduce breaking changes that existing parsers cannot handle.
+- `appended_offsets`: File offsets (0-based) for appended data.
+  If no data is appended, all offsets must be zero.
+  This can be used to reliably append data for logs that may stop in the middle of a message.
+
+  A process appending data should do:
+  - set the relevant `incompat_flags` bit,
+  - set the first `appended_offsets` that is 0 to the length of the log file,
+  - then append any type of messages that are valid for the Data section.
+
+It is possible that there are more fields appended at the end of this message in future ULog specifications.
+This means a parser must not assume a fixed length of this message.
+If the message is longer than expected (currently 40 bytes), the exceeding bytes must just be ignored.
+
+#### 'F': Format Definition Message
+
+Format definition that can be logged or used in another format definitions as a nested type.
+
+```
+struct message_format_s {
+  struct message_header_s header;
+  char format[header.msg_size];
+};
+```
+
+* `format` is a plain-text string with the following format: `message_name:field0;field1;`
+
+There can be an arbitrary amount of fields (minimum 1), separated by `;`.
+
+A field has the format: `type field_name` or `type[array_length] field_name` for arrays (only fixed size arrays are supported).
+
+`type` is one of the basic binary types or a `message_name` of another format definition (nested usage).
+A type can be used before it's defined.
+There can be arbitrary nesting but no circular dependencies.
+
+Some field names are special:
+- `timestamp`: every [Subscription Message](#a--subscription-message) must include a timestamp field
+  - Its type can be: `uint64_t` (currently the only one used), `uint32_t`, `uint16_t` or `uint8_t`.
+  - The unit is always microseconds, except for in `uint8_t` it's milliseconds.
+  - A log writer must make sure to log messages often enough to be able to detect wrap-arounds and a log reader must handle wrap-arounds (and take into account dropouts).
```

Comment:
@bkueng I don't understand this part. Could you explain?

Reply:
What it means is that the size of the timestamp isn't infinite, and it will rollover back to zero after a while. You can detect this is you log with some frequency, but if your log length is too infrequent you won't know if your latest item is actually at timestamp X or timestamp X + the interval for the whole timer to cycle once.

Reply:
Added to the doc!